### DISCUSSION TITLE
When key management is registering with OpenSSL and Sign&Verify registering with qatprovider, why both are offloading to OpenSSL?

### DISCUSSION THREAD

**Discussion by bjayanax** [openssl/openssl#24353] [Q&A]
When we registered keymgmt and sign&verify with qatprovider both are offloading to qatprovider. If we are using keymgmt from OpenSSL and sign&verify from qatprovider both are offloading to OpenSSL. May I know reason for this behaviour?

Do we need to use both keymgmt and sign&verify from same library?

Note: Checked few algorithms(RSA, ECDSA & SM2) and Observed  this behavior

**Comment by t8m** [comment]
What do you mean by "registering" here? You'll have to describe the application calls and configuration for us to help. Of course to be able to move keys among providers all of them has to have implemented a keymgmt (at least partially).

**Reply by bjayanax** [reply]
Hi Tomas,
Sorry for the uncleared explanation!!

I have provided basic explanation of registration and initialization of qatprovider at the end. And taking sm2 as an example here.

we have used sm2 signatures from qatprovider and not stored any sm2 keymgmt operation from qatprovider and completely used sm2 keymgmt from openssl . Unfortunately signature is not using from qatprovider and both keymgmt & signature using from openssl. 

**sm2 keymgmt:** https://github.com/intel/QAT_Engine/blob/ba2035c04e826018478fdb458ce51f545de076eb/qat_prov_init.c#L296  ( removed this line and checked for experimental purpose)

**sm2 signature:** https://github.com/intel/QAT_Engine/blob/ba2035c04e826018478fdb458ce51f545de076eb/qat_prov_init.c#L312

As you  explained, signature is offloading to qatprovider  when we are creating basic sm2 keymgmt API's in qatprovider and inside API's we called OpenSSL keymgmt API's using function pointers.

Basic explanation about qatprovider loading and initialization:
        1. Initializing qatprovider(3rd party provider) with OSSL_provider_init()
	2. Created qat_dispatch_table of type OSSL_DISPATCH and stored qat_query.
	3. Inside qat_query,  loaded provider with "OSSL_PROVIDER_load()"& setting default properties of qatprovider
Based on operation_id coming from OpenSSL speed we are storing(registering) supporter operations like keymgmt, keyexch, signature and ciphers etc.,
